### PR TITLE
Tester class draft and unit tests

### PR DESCRIPTION
**Pull request by LordB0red**
Focuses on (my presumed) methods of Tester. 

Assumes that UserBase from my Admin class is used for Tester. When UserManager is finalized, probably will need changes made, but throwing this up just to have groundwork for the methods.

### CODE REVIEW

File: PMSCore/Tester.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    /// <summary>
+    /// A Tester user who managages test cases and bug reporting.
+    /// </summary>
+    public class Tester : UserBase
+    {
+        
+
+        // Properties specific to a Tester
+        public List<string> TestCases;
+        public List<(string Description, string Severity)> BugReports;
```
Comment (ivanmartin91) [inline PMSCore/Tester.cs]: These two lines (12-13): public List<string> TestCases;
                                       public List<(string Description, string Severity)> BugReports;
                                       
TestCases and BugReports might need to be changed from public fields to properties with private setters to achieve encapsulation.

File: PMSCore/Tester.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+        /// <summary>
+        /// Adds a new test case to the tester's list of test cases.
+        /// </summary>
+        public void AddTestCase(string testCaseDescription)
+        {
+            if (!string.IsNullOrWhiteSpace(testCaseDescription))
+            {
+                TestCases.Add(testCaseDescription);
+                Console.WriteLine($"Test case added: {testCaseDescription}");
```
Comment (ivanmartin91) [inline PMSCore/Tester.cs]: line 45:  Console.WriteLine($"Test case added: {testCaseDescription}"); 

- Console.WriteLine should be replaced with _logger.LogInformation for consistent logging, and it's important that _logger is initialized in the constructor.

File: PMSCore/Tester.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        /// Method to view all test cases created by the tester.
+        /// </summary>
+        public void ViewTestCases()
+        {
+            Console.WriteLine("Test Cases:");
+            foreach (var testCase in TestCases)
+            {
+                Console.WriteLine($"- {testCase}");
+            }
+        }
```
Comment (ivanmartin91) [inline PMSCore/Tester.cs]: these lines:             
            
            Console.WriteLine("Test Cases:");
            foreach (var testCase in TestCases)
            {
                Console.WriteLine($"- {testCase}");
            }
        }

ViewTestCases() can be changed to return a list of test cases or a formatted string because this way it directly writes output to the console, and that's not good for testing.

File: PMSCore/Tester.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        /// <summary>
+        /// Submits a new bug report
+        /// </summary>
+        public void SubmitBugReport(string description, string severity)
+        {
+            if (!string.IsNullOrWhiteSpace(description) && !string.IsNullOrWhiteSpace(severity))
+            {
+                BugReports.Add((description, severity));
+                Console.WriteLine($"Bug report submitted: {description} (Severity: {severity})");
```
Comment (ivanmartin91) [inline PMSCore/Tester.cs]: Console.WriteLine($"Bug report submitted: {description} (Severity: {severity})");

Console.WriteLine should be replaced with _logger.LogInformation

File: PMSCore/Permission.cs [hunk 5]
```diff
@@ -10,6 +10,9 @@ public enum Permission
         EMPLOYEE = 2,
         MANAGER = 3,
         ADMIN = 4,
-        SUPERUSER = 5
+        SUPERUSER = 5,
```
Comment (ivanmartin91) [inline PMSCore/Permission.cs]: SUPERUSER = 5 is showing twice, one is a duplicate

File: PMSCore/AdminClass.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                Console.WriteLine("Username can't be empty");
+                return null;
+            }
+
+            if (permission == Permission.DEFAULT || permission == Permission.ADMIN)
+            {
+                Console.WriteLine("Invalid permission type for user creation.");
+                return null;
+            }
+
```
Comment (ivanmartin91) [inline PMSCore/AdminClass.cs]: for these lines:

if (permission == Permission.DEFAULT || permission == Permission.ADMIN)
            {
                Console.WriteLine("Invalid permission type for user creation.");
                return null;
            }
            
enum might be good to keep valid or invalid permissions for creating users

File: PMSCore/AdminClass.cs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            };
+
+            if (newUser == null)
+            {
+                Console.WriteLine("Failed to create user due to unsupported permission type.");
+                return null;
+            }
+
+            ManagedUsers.Add(newUser);
+            Console.WriteLine($"User {newUser.UserName} created with role {permission}.");
```
Comment (ivanmartin91) [inline PMSCore/AdminClass.cs]: Console.WriteLine should be replaced with _logger.LogInformation for consistent logging

File: PMSCore/AdminClass.cs [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        public bool DeleteUser(UserBase user)
+        {
+            if (ManagedUsers.Remove(user))
+            {
+                Console.WriteLine($"User {user.UserName} has been removed.");
+                return true;
+            }
+
+            Console.WriteLine("Specified user not found.");
+            return false;
```
Comment (ivanmartin91) [inline PMSCore/AdminClass.cs]: please replace Console.WriteLine with _logger.LogError

File: PMSCore/AdminClass.cs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        /// <summary>
+        /// Updates an existing user's info
+        /// </summary>
+        public bool UpdateUser(UserBase user, string newUserName, Permission newPermission)
+        {
+            if (!ManagedUsers.Contains(user))
+            {
+                Console.WriteLine("Specified user not found.");
```
Comment (ivanmartin91) [inline PMSCore/AdminClass.cs]: please replace Console.WriteLine with _logger.LogError

File: PMSCore/AdminClass.cs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+            if (!string.IsNullOrWhiteSpace(newUserName))
+            {
+                user.UserName = newUserName;
+                Console.WriteLine($"Updated username for user {user.UserID} to {newUserName}.");
+            }
+
+            if (Enum.IsDefined(typeof(Permission), newPermission) && newPermission != Permission.DEFAULT)
+            {
+                user.SetPermission(newPermission);
+                Console.WriteLine($"Updated permission for user {user.UserID} to {newPermission}.");
```
Comment (ivanmartin91) [inline PMSCore/AdminClass.cs]: please replace Console.WriteLine with _logger.LogInformation

File: PMSCore.Test/TesterTest.cs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+        [Fact]
+        public void AddTestCase_AddsAValidTestCaseToList()
+        {
+            // Arrange
+            var tester = new Tester("Tester1");
+
+            // Act
+            tester.AddTestCase("Test case 1");
+
```
Comment (ivanmartin91) [inline PMSCore.Test/TesterTest.cs]: Modify the Tester class to return a List<string> and validate the result directly in tests instead of relying on captured console output.

File: PMSCore.Test/TesterTest.cs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        [Fact]
+        public void SubmitBugReport_AValidReport_AddsToTheBugReports()
+        {
+            // Arrange
+            var tester = new Tester("Tester1");
+
+            // Act
+            tester.SubmitBugReport("Bug 1", "High");
+
```
Comment (ivanmartin91) [inline PMSCore.Test/TesterTest.cs]: in this line: tester.SubmitBugReport("Bug 1", "High");

change ViewBugReports to return a formatted string or list to optimize testing.

File: PMSCore.Test/TesterTest.cs
Comment (ivanmartin91) [inline PMSCore.Test/TesterTest.cs:1]: All the test methods should validate edge cases such as null or empty inputs explicitly.

File: PMSCore.Test/AdminClassTest.cs [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    {
+        [Fact]
+        public void CreateUser_ShouldCreateATester_WhenTesterPermissionIsApplied()
+        {
+            // Arrange - creating an Admin instance
+            var admin = new Admin("adminUser");
+
+            // Act - calling the CreateUser method
+            var user = admin.CreateUser("testerUser", Permission.TESTER);
+
```
Comment (ivanmartin91) [inline PMSCore.Test/AdminClassTest.cs]: for this line: var user = admin.CreateUser("testerUser", Permission.TESTER); 
you can add a test that would check for duplicate user prevention

File: PMSCore.Test/AdminClassTest.cs [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        [Fact]
+        public void UpdateUser_ShouldUpdateUsername_WhenInputIsValid()
+        {
+            // Arrange
+            var admin = new Admin("adminUser");
+            var user = admin.CreateUser("testerUser", Permission.TESTER);
+
+            // Act
+            var result = admin.UpdateUser(user, "updatedTester", Permission.TESTER);
+
```
Comment (ivanmartin91) [inline PMSCore.Test/AdminClassTest.cs]: for this line: var result = admin.UpdateUser(user, "updatedTester", Permission.TESTER);

please add a test to handle null inputs

File: PMSCore.Test/AdminClassTest.cs
Comment (ivanmartin91) [inline PMSCore.Test/AdminClassTest.cs:1]: Add more test cases for edge conditions, like:

- Trying to update or delete a user that doesn’t exist
- Verifying invalid permissions when using UpdateUser
